### PR TITLE
Higher Level Exceptions

### PR DESCRIPTION
**Pull request by hpohekar**
* Adding higher level custom exceptions in `src\ansys\fluent\core\solver\error_message.py`.

* Updating respective tests for added exceptions.

### CODE REVIEW

File: src/ansys/fluent/core/fluent_connection.py
Comment (seanpearsonuk) [inline src/ansys/fluent/core/fluent_connection.py:671]: :+1: for the inheritance.

**Review by seanpearsonuk — APPROVED** [review]
:+1: for starting with a single example. This is exactly what we wanted.

File: src/ansys/fluent/core/fluent_connection.py
Comment (hpohekar) [reply src/ansys/fluent/core/fluent_connection.py:671]: Right, Thank you.

File: src/ansys/fluent/core/file_session.py
Comment (mkundu1) [inline src/ansys/fluent/core/file_session.py:720]: At the consumer level, these exception objects are still of type TransactionError and the way to identify a SURFACE_NAMES_IDs_NOT_PROVIDED exception is to check for the error message. Do we want to refine the exception hierarchy further by deriving a SurfaceNotSpecified from TransactionError?

All the custom exception classes can be moved to a separate file, so user can find them in one place.

File: src/ansys/fluent/core/file_session.py
Comment (raph-luc) [inline src/ansys/fluent/core/file_session.py:112]: `TransactionError.SURFACE_NAMES_IDs_NOT_PROVIDED` is not a exception type, it can't be caught using `except`. 

I believe it makes more sense to list here just `TransactionError` as that is the exception type that is actually being raised and that can be caught.

File: src/ansys/fluent/core/fluent_connection.py
Comment (raph-luc) [inline src/ansys/fluent/core/fluent_connection.py:678]: I'm unsure about the usefulness of creating new exception class methods for exception messages that are only raised in one location in the code, as seems to be the case with `PORT_NOT_PROVIDED`.

Regardless, if this is how we would like to structure things moving forward, I understand.

File: src/ansys/fluent/core/file_session.py
Comment (hpohekar) [reply src/ansys/fluent/core/file_session.py:112]: Until now, we are not catching exceptions, therefore we have not derived `TransactionError` from exception class.

File: src/ansys/fluent/core/file_session.py
Comment (hpohekar) [reply src/ansys/fluent/core/file_session.py:720]: Discussed with @mkundu1, we will refine the exception hierarchy. Thank you.

File: src/ansys/fluent/core/file_session.py
Comment (hpohekar) [reply src/ansys/fluent/core/file_session.py:112]: @raph-luc I just came across following test code.

image

File: src/ansys/fluent/core/fluent_connection.py
Comment (seanpearsonuk) [reply src/ansys/fluent/core/fluent_connection.py:678]: @raph-luc Please allow me to explain the idea here via an example.  
`launch_fluent(...)` is called. Maybe I call it from my Python interpreter, or it's called from some other application (could be a cloud application). Fluent launches successfully but PyFluent fails to connect to Fluent. The calling application will catch exceptions raised by `launch_fluent`. It will use the exception information for observability and potentially for recovery. The exception type is the main source of information. Programmatically, that type is the best thing for deciding how to proceed. The message is useless in this respect if we are trying to do things rigorously. What are the useful exception types that can potentially help us to determine the recovery strategy? Does a FluentConnection error tell us anything more than a ValueError (or any other low-level exception)? Regarding the one location argument: if we can reuse exception types, that's good, but it's not a requirement.

File: src/ansys/fluent/core/file_session.py
Comment (raph-luc) [reply src/ansys/fluent/core/file_session.py:112]: Thanks @hpohekar, that is an example of exactly what I am saying: you can't catch `TransactionError.SURFACE_NAMES_IDs_NOT_PROVIDED`, but you can catch `TransactionError`. 

Change your except block to `except Error` and it will work. In this example, it is therefore the `Error` type that should be listed in the numpy-style docstring, in my understanding.

File: src/ansys/fluent/core/fluent_connection.py
Comment (raph-luc) [reply src/ansys/fluent/core/fluent_connection.py:678]: Thanks @seanpearsonuk, I understand that part, in this example `FluentConnectionError` can be caught, and that can therefore be a useful addition.

However, `FluentConnectionError.PORT_NOT_PROVIDED` is of little use in my understanding. It cannot be caught directly, it is just wrapping around the `"The port to connect to Fluent session is not provided."` message.

In this case, if `FluentConnectionError.PORT_NOT_PROVIDED` is only used in one location, then we are replacing what could be just `raise FluentConnectionError("The port to connect to Fluent session is not provided.")` in that location, by:
```
raise FluentConnectionError.PORT_NOT_PROVIDED()
```
plus the definition
```
    @classmethod
    def PROCESS_NOT_FINISHED(cls):
        return cls(
            "Fluent remote instance not supported by FluentConnection.wait_process_finished()."
        )
```
Which doesn't seem very efficient, to me at least. 

Am I misunderstanding or missing something?

File: src/ansys/fluent/core/file_session.py
Comment (seanpearsonuk) [reply src/ansys/fluent/core/file_session.py:112]: @raph-luc judging by @hpohekar 's other comments below, this issue should resolve itself shortly.

File: src/ansys/fluent/core/file_session.py
Comment (seanpearsonuk) [reply src/ansys/fluent/core/file_session.py:720]: @hpohekar you could also look into refining exception hierarchy in the units repo if it interests you - I believe that what we are doing there is similar to what we currently see here.

File: src/ansys/fluent/core/fluent_connection.py
Comment (seanpearsonuk) [reply src/ansys/fluent/core/fluent_connection.py:678]: @raph-luc No, you are not misunderstanding. There are just some mistakes in the implementation which @hpohekar is working through. Thanks for notifying him about that.

Comment (hpohekar) [reply src/ansys/fluent/core/fluent_connection.py:678]: @raph-luc The suggestion given by @mkundu1 to separate out class methods as hierarchy will resolve your concern mentioned above.

File: src/ansys/fluent/core/file_session.py
Comment (hpohekar) [reply src/ansys/fluent/core/file_session.py:720]: Yes, after completing this task I will look into it.

**Comment by seanpearsonuk** [comment]
@hpohekar You might have seen my early comment stating that the fact this PR represents a small change is a good thing. 

Let's keep these changes focussed in a small area. Let's focus on `launch_fluent` first. It's a great example of a function that's composed of multiple steps that can go wrong.

Write tests that demonstrate that we can catch distinct custom exceptions for each failure type, including (in the first instance) _failed to start fluent_ and _failed to connect to fluent_.

File: src/ansys/fluent/core/file_session.py
Comment (mkundu1) [reply src/ansys/fluent/core/file_session.py:720]: @hpohekar @seanpearsonuk Another usage pattern could be:
```
try:
    <pyfluent code>
except TransactionError as ex:
    if ex.code == TransactionError.ErrorCode.SURFACE_NAMES_IDS_NOT_PROVIDED:
        <handle SURFACE_NAMES_IDS_NOT_PROVIDED exception>
    elif ex.code == TransactionError.ErrorCode.INVALID_FIELD_NAME_PREFIX:
        <handle INVALID_FIELD_NAME_PREFIX exception>
    else:
        <handle other TransactionError>
```
`ErrorCode` is an enum. This way we won't have too many custom exception classes, but users will be able to identify the fine-grained errors.

Comment (raph-luc) [reply src/ansys/fluent/core/file_session.py:720]: I personally would vote for using and catching specific exception types if possible, like `SurfaceNotSpecifiedError` as you said earlier, otherwise that structure you just shared is going back to being very close to just parsing the Exception error message as in this comment and my comment after that: https://github.com/ansys/pyfluent/issues/2098#issuecomment-1750759704

Which doesn't require any custom exception classes.

Comment (raph-luc) [reply src/ansys/fluent/core/file_session.py:720]: They do have this:
image
Which is closer to what I believe we should be aiming for if we are going to add more custom exception classes:

Comment (mkundu1) [reply src/ansys/fluent/core/file_session.py:720]: @raph-luc There is a lack of official documentation here. [This](https://github.com/ansys/pyfluent-reader/pull/14#pullrequestreview-1688178548) is the general usage pattern for handling exceptions from a Python client.

Right, internally they have implemented a full custom class hierarchy, but that implementation is private.

Comment (raph-luc) [reply src/ansys/fluent/core/file_session.py:720]: I see, thanks for the explanation. As they have those status codes which are used repeatedly and relevant everywhere in their modules (see https://grpc.github.io/grpc/core/md_doc_statuscodes.html), it makes sense that they would include the status codes as part of their custom exception classes.

Comment (hpohekar) [reply src/ansys/fluent/core/file_session.py:720]: @mkundu1 This looks good to me. Thank you.

image

File: tests/test_launcher.py
Comment (seanpearsonuk) [inline tests/test_launcher.py:91]: @hpohekar It's kind of remarkable that the exception types are different. That's the great thing about this kind of testing. Please continue to add more scenarios, and as you go, you can review if the raised exceptions match the error scenarios (and update the the tests and implementations as necessary). Thanks.

Comment (hpohekar) [reply tests/test_launcher.py:91]: Yes sure.

File: src/ansys/fluent/core/meta.py
Comment (raph-luc) [inline src/ansys/fluent/core/meta.py:747]: Are we going to upgrade this `DISALLOWED_VALUES` to an exception class as well so that it can be caught directly (similarly to what was done for `FluentConnectionError.PortNotProvidedError` and the others)?

File: src/ansys/fluent/core/solver/error_message.py
Comment (raph-luc) [inline src/ansys/fluent/core/solver/error_message.py:1]: Would a file name like `exceptions.py` rather than `error_message.py` work better?

So that we can, for example,
`from ansys.fluent.core.solver.exceptions import TransactionError`
rather than
`from ansys.fluent.core.solver.error_message import TransactionError`

**Review by raph-luc — COMMENTED** [review]
Looks great, thanks @hpohekar, only a couple minor nitpicks below

Will wait to review again and approve when you set this as ready for review again

File: src/ansys/fluent/core/solver/error_message.py
Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py:1]: Let's see what others think. Personally I just wanted to incorporate this change without disturbing current use and structure of `error_message.py`, therefore I haven't renamed it yet.

File: src/ansys/fluent/core/meta.py
Comment (hpohekar) [reply src/ansys/fluent/core/meta.py:747]: We haven't updated it with latest change yet. We will do it soon.

File: src/ansys/fluent/core/solver/error_message.py
Comment (seanpearsonuk) [reply src/ansys/fluent/core/solver/error_message.py:1]: If the name is no longer accurate, it shall be changed.

File: tests/test_launcher.py
Comment (seanpearsonuk) [reply tests/test_launcher.py:91]: @hpohekar we should keep the comment unresolved while the points are still being addressed. Let me specify what they are. I was referring to the arbitrary exception type differences. The original `launch_fluent` code here was wrong. Both exceptions should have been `TypeError`, and that has led you to create distinct types, which looks very odd in this test, right? We should ask about the value of adding new exception types in each case. These `TypeError`s will crop up do to incorrect usage of the interface. `TypeError` is already adequate for that, and the message contains all the context. In creating new exception types, we should focus on the more sophisticated failure scenarios.

File: src/ansys/fluent/core/solver/error_message.py
Comment (seanpearsonuk) [inline src/ansys/fluent/core/solver/error_message.py:30]: Why does this inherit from `ValueError`?

File: src/ansys/fluent/core/solver/error_message.py
Comment (seanpearsonuk) [inline src/ansys/fluent/core/solver/error_message.py:50]: Why does this inherit from `RuntimeError`?

File: src/ansys/fluent/core/solver/error_message.py
Comment (seanpearsonuk) [inline src/ansys/fluent/core/solver/error_message.py:69]: Something not provided sounds like a `TypeError`. CHeck if the new type is really needed. If you focus on testing then we can see what is truly needed.

File: src/ansys/fluent/core/solver/error_message.py
Comment (seanpearsonuk) [reply src/ansys/fluent/core/solver/error_message.py:1]: ...although if we have XError directly associated with class x.X then XError should probably be in module x as well. That would also mean that we can eliminate all this extra namespacing which looks heavyweight.

File: src/ansys/fluent/core/solver/error_message.py
Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py:30]: image

https://stackoverflow.com/questions/39077623/can-a-python-inner-class-be-a-subclass-of-its-own-outer-class#:~:text=not%20a%20subclass.-,No%2C%20an%20inner%20class%20can't%20inherit%20(not%20extend,while%20defining%20the%20inner%20class.

File: src/ansys/fluent/core/solver/error_message.py
Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py:69]: Yes, there are many places where incorrect exceptions has been raised. I'm not focusing on correcting each of them right now, I will first ensure that all of them have been covered as per new exception format then after passing all tests, I will correct those exceptions one by one.

File: tests/test_launcher.py
Comment (hpohekar) [reply tests/test_launcher.py:91]: @seanpearsonuk 

`The original launch_fluent code here was wrong. Both exceptions should have been TypeError, and that has led you to create distinct types, which looks very odd in this test, right?` 

Even if we had corrected both of them to `TypeError` still we need to create and inherit distinctly as we have done in `TransactionError`.

`We should ask about the value of adding new exception types in each case.`

The issue description says that we don't want lower-level built-in exception types, therefore we are updating all existing exceptions. We have not decided any criteria to add new exception types (places where we want to add new exceptions types and places where we want to use built-in exception types).

File: src/ansys/fluent/core/solver/error_message.py
Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py:1]: shall we move all new exceptions to respective modules ? It will also keep module documentation consistent with the new exception type usage.

File: src/ansys/fluent/core/solver/error_message.py
Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py]: Right, we followed `TransactionError` format which includes all `RuntimeError`. it's removed now. Thank you.

File: src/ansys/fluent/core/solver/error_message.py
Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py]: It's removed now. Thank you.

File: src/ansys/fluent/core/file_session.py
Comment (hpohekar) [inline src/ansys/fluent/core/file_session.py:113]: Both of them should be `TypeError`

File: src/ansys/fluent/core/file_session.py
Comment (hpohekar) [inline src/ansys/fluent/core/file_session.py:111]: Should be TypeError

File: src/ansys/fluent/core/file_session.py
Comment (hpohekar) [inline src/ansys/fluent/core/file_session.py:230]: Should be ValueError

File: src/ansys/fluent/core/launcher/launcher.py
Comment (hpohekar) [inline src/ansys/fluent/core/launcher/launcher.py:606]: Should be ValueError

File: src/ansys/fluent/core/solver/error_message.py
Comment (raph-luc) [reply src/ansys/fluent/core/solver/error_message.py:1]: My vote would be for leaving `error_message.py` as it is (before these changes), and either:
- Adding a new `exceptions.py` module for all the PyFluent exception classes (assuming they are shared by different modules and each exception class isn't used only in a single module) or 
- As Sean suggested, adding each Exception class to the module where it is used (if each exception class is used in one location)

File: src/ansys/fluent/core/meta.py
Comment (hpohekar) [reply src/ansys/fluent/core/meta.py]: Done.

File: tests/test_launcher.py
Comment (hpohekar) [reply tests/test_launcher.py:91]: We can replace `LauncherError.MeshingModeError` with `LauncherError.UnexpectedKeywordArgumentError`.

We can also remove `LauncherError.MeshingModeError`.

Comment (hpohekar) [reply tests/test_launcher.py]: Replaced `LauncherError.MeshingModeError` with `LauncherError.UnexpectedKeywordArgumentError`

File: src/ansys/fluent/core/solver/error_message.py
Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py:73]: @seanpearsonuk This should be `ValueError` not `TypeError`.

"`TypeError` is raised when an operation or function is applied to an object of inappropriate type. The associated value is a string giving details about the type mismatch."

Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py]: Corrected error type.

File: src/ansys/fluent/core/solver/error_message.py
Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py:1]: @seanpearsonuk @raph-luc 

We have `error_message.py` at 2 places.

1. src/ansys/fluent/core/solver/error_message.py
2. fluent/cortex/pylib/flapi/error_message.py

Therefore if we are going to add new `exceptions.py` then we will have to add it at above 2 mentioned places.

I think we should follow only 1 approach out of following - 

1. Keep all new exceptions in `error_message.py` and we can rename it to `exceptions.py`. 
(It is useful as `DisallowedValuesError` is defined in single file and we are using it in various other files) 

OR

2.  Keep all new exceptions in respective modules. 
(We will have to import various common errors from various files e.g. `DisallowedValuesError`)

I will go with 1st approach.

Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py:1]: @raph-luc 

"Edit: also this exceptions.py module probably should be under ansys.fluent.core and not ansys.fluent.core.solver as it is not solver-specific and is for PyFluent in general, right?"

In Fluent we have `error_message.py` if we leave it as it is, then we will have to add and maintain `exceptions.py` there additionally and I think it's not useful. 

We have renamed `error_message.py` to `exceptions.py` and we are keeping it with `flobject.py` to be consistent with Fluent.

Comment (seanpearsonuk) [reply src/ansys/fluent/core/solver/error_message.py:1]: Re:1 error_message.py contains functions that build similar error messages (i.e. currently messages, not exception types) for different contexts. That was done simply to avoid code duplication. It has had to be a shared module because of the wide range of usage contexts. If something is associated with one module, be it a class, function or their associated exception types, then it belongs in that module, unless a strong counterargument can be provided.

Comment (seanpearsonuk) [reply src/ansys/fluent/core/solver/error_message.py:1]: > @raph-luc
> 
> "Edit: also this exceptions.py module probably should be under ansys.fluent.core and not ansys.fluent.core.solver as it is not solver-specific and is for PyFluent in general, right?"
> 
> In Fluent we have `error_message.py` if we leave it as it is, then we will have to add and maintain `exceptions.py` there additionally and I think it's not useful.
> 
> We have renamed `error_message.py` to `exceptions.py` and we are keeping it with `flobject.py` to be consistent with Fluent.

You need to start resisting the urge to bundle unrelated things together.

File: src/ansys/fluent/core/file_session.py
Comment (seanpearsonuk) [inline src/ansys/fluent/core/file_session.py:112]: `TransactionError` is being used to provide a namespace, but the module already provides a namespace. Also, `file_session` would addsome meaning, whereas `TransactionError` is generic.

File: src/ansys/fluent/core/solver/error_message.py
Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py:1]: > > @raph-luc
> > "Edit: also this exceptions.py module probably should be under ansys.fluent.core and not ansys.fluent.core.solver as it is not solver-specific and is for PyFluent in general, right?"
> > In Fluent we have `error_message.py` if we leave it as it is, then we will have to add and maintain `exceptions.py` there additionally and I think it's not useful.
> > We have renamed `error_message.py` to `exceptions.py` and we are keeping it with `flobject.py` to be consistent with Fluent.
> 
> You need to start resisting the urge to bundle unrelated things together.

@seanpearsonuk 

I don't think that above point is unrelated. let me clarify myself.

1. First we decided to rename `error_message.py` to `exceptions.py` because we are adding exceptions to it now.

image

2. Then we have decided to have XError which is directly associated to x.X in the module as well to avoid extra namespacing.

image

Related to this I asked shall we move all new exception types to respective module ? - unanswered yet.

3. I found that we have added a new exception type in flobject.py.

image

So there are following scenarios right now.

1. Add `InactiveObjectError` in flobject.py itself. (Haven't decided yet)

2. Add `InactiveObjectError` in `error_message.py` which is renamed to `exceptions.py`. (Done in current PR)

3. Add new `exceptions.py` in PyFluent core and in Fluent. (Not done yet)

I just thought adding `InactiveObjectError` in `error_message.py` which is renamed to `exceptions.py` is good option from point of code duplication and maintenance.

Could you please explain how above viewpoint is unrelated to the change I made, regarding renaming `error_message.py` to `exception.py` and adding `InactiveObjectError` to it  to be consistent with Fluent ?

Thank you.

Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py:1]: > Re:1 error_message.py contains functions that build similar error messages (i.e. currently messages, not exception types) for different contexts. That was done simply to avoid code duplication. It has had to be a shared module because of the wide range of usage contexts. If something is associated with one module OTOH, be it a class, function or their associated exception types, then it belongs in that module, unless a strong counterargument can be provided otherwise.

@seanpearsonuk 

Yes I agree. Therefore I have added all of them to respective modules in first commit right ? Then we got suggestions that we should maintain all exceptions in a single file therefore I started adding them in `error_message.py`.

Comment (seanpearsonuk) [reply src/ansys/fluent/core/solver/error_message.py:1]: OK, by your description, we are converging on the same thing. Sounds good.

Comment (hpohekar) [reply src/ansys/fluent/core/solver/error_message.py:1]: So from discussion regarding this, we are going to do following changes.

1. Add exception.py at 
    a) ansys/fluent/core/exception.py
    b) fluent/cortex/pylib/flapi/exception.py
    
 2. Keep `error_message.py` as it is in both PyFluent and Fluent.
 
 3. Keep common exceptions in `exceptions.py`
 
 4. Move distinct exception types belonging to single class into respective modules.

**Comment by hpohekar** [comment]
https://github.com/ansys/pyfluent/pull/2176

**Pull request closed without merging**
